### ISSUE TITLE
`rr_ps_ns` sometimes times out

### ISSUE THREAD

**Issue by rocallahan** [mozilla/rr#1978]
On my machine. It looks like the test hangs during recording. It's not easily reproducible.

**Comment by Keno** [comment]
Just saw this one as well.

**Comment by Keno** [comment]
I'm wondering if we should change the test infrastructure to take some more info on timed-out test cases. E.g. `/proc/<x>/stack`, `/proc/<x>/status` and a backtrace of rr itself.

**Comment by rocallahan** [comment]
Yes, and we should probably do that for rr assertion failures too.

**Comment by rocallahan** [comment]
I don't see how to get ctest to do something special on a timeout. We may need to raise the ctest timeout and introduce a timeout mechanism in util.sh that triggers earlier.

**Comment by rocallahan** [comment]
I've done this: https://github.com/mozilla/rr/commit/63cb183c96cf4475c15ac72b89a4226a6e15eafc
More detailed update mailed to rr-dev.
Hopefully we'll have a lot more useful data now when a test fails, especially with a timeout.

**Comment by Keno** [comment]
https://gist.github.com/Keno/936491e9002480b5974c5e02f6b88ac3

**Comment by rocallahan** [comment]
Thanks!

According to this, the tracee's child process was already created *and exited*, but `prepare_clone` is in the path that says
````
      // clone failed and we're existing the syscall with an error. Reenter
      // the syscall so that we're in the same state as the normal execution
      // path.
````

**Comment by rocallahan** [comment]
And it looks like the tracee process is in a signal-delivery-stop.

**Comment by rocallahan** [comment]
I assume this is a kernel where `try_setup_ns(CLONE_NEWPID)` *doesn't* fork?

**Comment by rocallahan** [comment]
Interestingly, rr is running when we stop, so maybe somehow `Task::enter_syscall` is in an infinite loop...

**Comment by rocallahan** [comment]
> I assume this is a kernel where try_setup_ns(CLONE_NEWPID) doesn't fork?

No, I guess it does.

**Comment by rocallahan** [comment]
`rr_ps_ns` actually fails when run under gdb or strace, for me.

**Comment by rocallahan** [comment]
Hmm, I see that `try_setup_ns_internal` doesn't wait to reap its child. It should. But I don't actually understand how it works now... Shouldn't we fork the child *after* `unshare`? Why is that procfile munging in a child  process anyway?

**Comment by rocallahan** [comment]
https://github.com/mozilla/rr/commit/a771deeb7dcef01f4ac353a4215c3b00911fa919 and https://github.com/mozilla/rr/commit/28be3fcd8eefd405c4da4b8c133066f4b3a2013e may shed a little light. But we should add a `waitpid` to nsutils.h if I knew where...

**Comment by rocallahan** [comment]
This is what I get from strace:
````
clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7ff59c0e8150) = ? ERESTARTNOINTR (To be restarted)
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=26324, si_uid=0, si_status=0, si_utime=0, si_stime=0} ---
clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7ff59c0e8150) = -1 ENOMEM (Cannot allocate memory)
````
This is the clone being interrupted by the SIGCHLD from the exit of the previously forked helper process.

**Comment by Keno** [comment]
> Why is that procfile munging in a child process anyway?

In general, it needs to be done from a process that's in the original namespace, so we need to fork one off before we unshare. I'll submit a PR to add a proper waitpid.

**Comment by Keno** [comment]
As far as I'm aware you need to write those files after unsharing the user namespace, so either you do it in a child process, or you fork and have that process be the one that returns back out of the function. The latter alternative seemed more confusing.

**Comment by rocallahan** [comment]
Right, sorry.
